src/providers/aws.js

```diff
@@ -0,0 +1,88 @@
+const AbstractFileTransfer = require('ms-files-transport');
+const S3 = require('aws-sdk/clients/s3');
+const { merge } = require('lodash');
+const assert = require('assert');
+
+const UPLOAD_URL_EXPIRES_IN_SEC = 60000;
+const DOWNLOAD_URL_EXPIRES_IN_SEC = 60000;
+
+class AWSTransport extends AbstractFileTransfer {
+  constructor(opts = {}) {
+    super();
+
+    const config = merge({}, AWSTransport.defaultOpts, opts);
+    this._config = config;
+    this._logger = this._config.logger;
+
+    assert(config !== undefined);
+    assert(config.options !== undefined);
+    assert(config.options.accessKeyId !== undefined);
+    assert(config.options.secretAccessKey !== undefined);
+    assert(config.options.region !== undefined);
+    assert(config.options.bucketName !== undefined);
+
+    this.setupS3();
+  }
+
+  setupS3() {
+    try {
+      this._s3 = new S3({
+        signatureVersion: 'v4',
+        region: this._config.region,
+        credentials: {
+          accessKeyId: this._config.accessKeyId,
+          secretAccessKey: this._config.secretAccessKey,
+        },
+      });
+    } catch (e) {
+      this._logger.warn('failed to load aws-sdk/clients/s3', e);
+    }
+  }
+
+  connect() {
+    return Promise
+      .bind(this)
+      .then(this.findOrCreateBucket);
+  }
+
+  async findOrCreateBucket() {
+    const { bucketName } = this._config;
+
+    const s3 = this._s3;
+
+    const buckets = s3.listBuckets();
+    const exists = buckets.find(bucketName);
+
+    if (!exists) {
+      this.log.debug('initiating createBucket: %s', bucketName);
+      await this._s3.createBucket({
+        Bucket: bucketName,
+      });
+    }
+  }
+
+  async getSignedUrl(action, input) {
+    const params = {
+      Bucket: this._config.bucketName,
+      Expires: action === 'putObject' ? UPLOAD_URL_EXPIRES_IN_SEC : DOWNLOAD_URL_EXPIRES_IN_SEC,
+      Key: input.key,
+    };
```

Comment:
not familiar with s3 signed URLs, but on google you have to sign headers as well, especially for put action.
headers include object size, any custom headers, etc - could this be required here?